drivers/ethernet/eth_virtio_net.c

```diff
@@ -0,0 +1,254 @@
+	       ETHERNET_LINK_2500BASE | ETHERNET_LINK_5000BASE;
+}
+
+static int virtnet_send(const struct device *dev, struct net_pkt *pkt)
+{
+	const struct virtnet_config *config = dev->config;
+	struct virtnet_data *data = dev->data;
+	size_t len = net_pkt_get_len(pkt);
+
+	if (net_pkt_read(pkt, data->txb + sizeof(struct _virtio_net_hdr), len)) {
+		LOG_ERR("could not read contents of packet to be sent");
+		return -EIO;
+	}
+
+	struct virtq *vq = virtio_get_virtqueue(config->vdev, VIRTQ_TX(1));
+	struct virtq_buf vqbuf[] = {
+		{.addr = data->txb, .len = sizeof(struct _virtio_net_hdr) + len}};
+
+	if (virtq_add_buffer_chain(vq, vqbuf, 1, 1, NULL, NULL, K_FOREVER)) {
+		LOG_ERR("could not send packet");
+		return -EIO;
+	}
+	virtio_notify_virtqueue(config->vdev, VIRTQ_TX(1));
+	return 0;
+}
+
+void virtnet_rx_cb(void *priv, uint32_t len)
+{
+	const struct _rx_cb_data *p = priv;
+	struct virtnet_data *data = p->data;
+	uint16_t buf_no = p->buf_no;
+	const struct virtnet_config *config = data->dev->config;
+	struct virtq *vq = virtio_get_virtqueue(config->vdev, VIRTQ_RX(1));
+
+	len -= sizeof(struct _virtio_net_hdr);
+	struct net_pkt *pkt =
+		net_pkt_rx_alloc_with_buffer(data->iface, len, AF_UNSPEC, 0, K_FOREVER);
+
+	if (pkt == NULL) {
+		LOG_ERR("received packet, but could not pass it to the operating system");
+	} else if (net_pkt_write(pkt, &(data->rxb[buf_no][sizeof(struct _virtio_net_hdr)]), len)) {
+		LOG_ERR("could not copy entire received packet");
+		net_pkt_unref(pkt);
+	} else if (net_recv_data(data->iface, pkt)) {
+		LOG_ERR("operating system failed to receive packet");
+		net_pkt_unref(pkt);
+	}
+	struct virtq_buf vqbuf[] = {{.addr = &(data->rxb[buf_no]), .len = VIRTIO_NET_BUFLEN}};
+
+	virtq_add_buffer_chain(vq, vqbuf, 1, 0, virtnet_rx_cb, priv, K_FOREVER);
+	virtio_notify_virtqueue(config->vdev, VIRTQ_RX(1));
+}
+
+static void virtnet_if_init(struct net_if *iface)
+{
+	ethernet_init(iface);
+	const struct device *dev = net_if_get_device(iface);
+	struct virtnet_data *data = dev->data;
+	const struct virtnet_config *config = dev->config;
+
+	if (dev == NULL) {
+		LOG_ERR("could not access device structure!");
+		return;
+	}
+	data->iface = iface;
+	net_if_set_link_addr(iface, data->mac, sizeof(data->virtio_devcfg->mac), NET_LINK_ETHERNET);
+	struct virtq *vq = virtio_get_virtqueue(config->vdev, VIRTQ_RX(1));
+
+	for (int i = 0; i < CONFIG_ETH_VIRTIO_NET_RX_BUFFERS; i++) {
+		data->rx_cb_data[i].data = data;
+		data->rx_cb_data[i].buf_no = i;
+
+		struct virtq_buf vqbuf[] = {{.addr = &(data->rxb[i]), .len = VIRTIO_NET_BUFLEN}};
+
+		virtq_add_buffer_chain(vq, vqbuf, 1, 0, virtnet_rx_cb, &(data->rx_cb_data[i]),
+				       K_FOREVER);
+		virtio_notify_virtqueue(config->vdev, VIRTQ_RX(1));
+	}
+	LOG_DBG("initialization finished");
+}
+
+static int virtnet_dev_init(const struct device *dev)
+{
+	const struct virtnet_config *config = dev->config;
+	struct virtnet_data *data = dev->data;
+
+#if DT_INST_PROP(0, zephyr_random_mac_address)
```

Comment:
this driver is instance based, access these values in VIRTIO_NET_DEFINE

local_mac_address can directly go in the data struct, zephyr_random_mac_address should be a bool in config, that is checked here